### PR TITLE
Implement API create World

### PR DESCRIPTION
**Pull request by odkhang**
As part of issue https://github.com/fossasia/eventyay-tickets/issues/358
This PR is implement API create world which using by eventyay-common

## Summary by Sourcery

Implement a new API endpoint to create or update a 'World' entity, including permission checks and JWT-based authentication. Update the settings to include a new 'DOMAIN_PATH' configuration for constructing world URLs.

New Features:
- Introduce a new API endpoint for creating or updating a 'World' entity, allowing users with the appropriate permissions to manage world configurations.

Enhancements:
- Add a new configuration setting 'DOMAIN_PATH' in the settings file to define the base domain path for world URLs.

### CODE REVIEW

**Review by mariobehling — CHANGES_REQUESTED** [review]
All server style tests are failing. Please fix.

File: server/venueless/api/views.py [hunk 1]
```diff
@@ -115,6 +119,90 @@ def get(self, request, **kwargs):
...
+                "JWT_secrets": [
+                    {
+                        "issuer": "any",
+                        "audience": "venueless",
+                        "secret": secret,
+                    }
+                ]
+            }
+
+            title_default = {
```
Comment (hongquan) [inline server/venueless/api/views.py:143]: This impementation leave to many holes:
- `data.get()` can return `None` and `None.items()` will throw exception.
- `set.pop()` will throw exception if `set` is empty.

Also, if you don't use `key`, retrieving values from dict instead.

File: server/venueless/api/views.py [hunk 2]
```diff
@@ -115,6 +119,90 @@ def get(self, request, **kwargs):
...
+                value for key, value in request.data.get("title").items() if value
+            }.pop()
+
+            # if world already exists, update it, otherwise create a new world
+            try:
+                if World.objects.filter(id=request.data.get("id")).exists():
```
Comment (hongquan) [inline server/venueless/api/views.py:149]: Data received from outside work (form, HTTP request) have to be validated.

You are calling 
```py
World.objects.get(id=request.data.get("id")
```

but what if some attackers pass "id=abc"? This code will throw exception.

File: server/venueless/api/views.py [hunk 3]
```diff
@@ -115,6 +119,70 @@
...
+                else:
+                    world = World.objects.create(
+                        id=request.data.get('id'),
+                        title=request.data.get('title')[request.data.get('locale')] or request.data.get('title')['en'],
+                        domain='{}/video/{}'.format(settings.DOMAIN_PATH, settings.BASE_PATH, request.data.get('id')),
+                        locale=request.data.get('locale'),
+                        timezone=request.data.get('timezone'),
+                        config=config,
+                    )
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (hongquan) [reply server/venueless/api/views.py:177]: This is a bad practice to catch a very general exception class. This has been reminded before.

File: server/venueless/api/views.py [hunk 4]
```diff
@@ -115,6 +119,70 @@
...
+        if auth_header and auth_header[0].lower() == b"bearer":
+            if len(auth_header) == 1:
+                raise exceptions.AuthenticationFailed(
+                    "Invalid token header. No credentials provided."
+                )
+            elif len(auth_header) > 2:
+                raise exceptions.AuthenticationFailed(
+                    "Invalid token header. Token string should not contain spaces."
+                )
+        payload = jwt.decode(auth_header[1], settings.SECRET_KEY, algorithms=["HS256"])
```
[the remark this answers is not in the document]
Comment (hongquan) [reply server/venueless/api/views.py:202]: Please follow sourcery-api.

File: server/venueless/api/views.py [hunk 5]
```diff
@@ -115,6 +119,90 @@ def get(self, request, **kwargs):
...
+                        settings.DOMAIN_PATH, settings.BASE_PATH, request.data.get("id")
+                    )
+                    world.locale = request.data.get("locale")
+                    world.timezone = request.data.get("timezone")
+                    world.save()
+                else:
+                    world = World.objects.create(
+                        id=request.data.get("id"),
+                        title=request.data.get("title")[request.data.get("locale")]
+                        or request.data.get("title")["en"]
```
Comment (hongquan) [inline server/venueless/api/views.py:166]: What if `data(get('title'))` returns `None` or something not a `dict`. Validation needs to be applied to make sure the incoming data has the expected data type and shape.

**Review by hongquan — CHANGES_REQUESTED** [review]

**Review by hongquan — APPROVED** [review]

**Review by mariobehling — APPROVED** [review]

**Pull request merged**